### PR TITLE
[Fork] "Do you want to fork this repository?" Dialog

### LINKED ISSUE #8548: [Fork] After cloning without write access, fork and publish from Desktop
_Extracted from https://github.com/desktop/desktop/issues/8509 (Problem 1) for tracking work on this particular issue_

**Problem: Trying to push changes to a repo you don't have write access to**

If you've cloned a repo and make some changes with the intent to contribute, you're basically stuck. You can't push those changes to the repo you've cloned because you don't have write access, and you can't bring them over to a fork that doesn't yet exist in a low-friction way.

Desktop could let you know you don't have write access and give you a nice pathway to fork and create a PR. This is conceptually similar to local commits to protected branches, where ideally we wouldn't allow you to do that (because you'd never be able to push those commits to that branch), and instead give you a nice path to somewhere where your work can eventually be pushed without issue.

A potential solution is outlined in https://github.com/desktop/desktop/issues/3918 by @say25.

### PR DESCRIPTION
**Pull request by outofambit**
closes #8548 

This PR adds the basic "make a fork" dialog flow, triggered by a failed push to a repo that the user doesn't have write access to.

**Create Fork Prompt**
Screen Shot 2020-01-17 at 4 45 00 PM

**Error Dialog**
Screen Shot 2020-01-17 at 4 53 36 PM

**Happy Flow**
fork-yay

## Description

* Adds error handler for push rejections due to missing write access to a repo
* Adds dialog for creating a fork
* Adds a wrapper for creating a fork via the GitHub API
* Adds `AppStore` method for redirecting a local repo to point to a fork

### CODE REVIEW

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      )
+      this.setState({ loading: false })
+      this.props.onDismissed()
+    } catch (e) {
+      log.error(`Fork creation through API failed (${e})`)
+      sendNonFatalException('forkCreation', e)
+      this.setState({ error: e })
+      this.setState({ loading: false })
+    }
+  }
```
Comment (outofambit) [inline app/src/ui/forks/create-fork-dialog.tsx]: @niik this is the critical logic flow in this PR!

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -5523,6 +5524,47 @@ export class AppStore extends TypedBaseStore<IAppState> {
...
+      repository,
+      account,
+    })
+  }
+
+  /**
+   * Converts a local repository to use the given fork
+   * as its default remote and associated `GitHubRepository`.
+   */
+  public async _convertRepositoryToFork(
```
Comment (niik) [inline app/src/lib/stores/app-store.ts]: Shouldn't this also add the current default remote as we do when cloning a regular fork? See `addUpstreamRemoteIfNeeded` and `updateExistingUpstreamRemote`. I suspect it'll eventually get added automatically but we should set it up correctly from the start.

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        this.props.repository,
+        fork
+      )
+      this.setState({ loading: false })
+      this.props.onDismissed()
+    } catch (e) {
+      log.error(`Fork creation through API failed (${e})`)
+      sendNonFatalException('forkCreation', e)
+      this.setState({ error: e })
+      this.setState({ loading: false })
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/forks/create-fork-dialog.tsx]: lol, why didn't i see that? 3667bf3d1

File: app/src/ui/dispatcher/error-handlers.ts [hunk 4]
```diff
@@ -580,6 +584,52 @@ export async function samlReauthRequired(error: Error, dispatcher: Dispatcher) {
...
+
+  if (!isRepositoryWithGitHubRepository(repository)) {
+    return error
+  }
+
+  if (hasWritePermission(repository.gitHubRepository)) {
+    return error
+  }
+
+  if (!pushFailureErrorTypes.has(dugiteError)) {
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/dispatcher/error-handlers.ts]: oh thanks, i didn't notice that helper! 3017c7526

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+function renderError(
+  repository: RepositoryWithGitHubRepository,
+  error: Error,
+  onDismissed: () => void
+) {
+  const suggestion =
+    repository.gitHubRepository.htmlURL !== null ? (
+      <>
+        You can try creating the fork manually at{' '}
+        <LinkButton>{repository.gitHubRepository.htmlURL}</LinkButton>.
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/forks/create-fork-dialog.tsx]: good catch and i like that alternative a lot! bfca1daa16be486644fdc69b75f42806131e1607

File: app/src/lib/stores/app-store.ts (hunk 2 above)
Comment (outofambit) [reply app/src/lib/stores/app-store.ts]: ah yes, thank you! 5446528da

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      return (
+        <Dialog
+          onDismissed={this.props.onDismissed}
+          onSubmit={this.onSubmit}
+          type="normal"
+          key={this.props.repository.name}
+          id="create-fork"
+        >
+          <DialogHeader
+            title="Do you want to fork this repository?"
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/forks/create-fork-dialog.tsx]: done in 37c6d1ac1

**Comment by billygriffin** [comment]
> One question: once they fork, are we okay with swapping out the upstream repo for the forked one without informing the user? Maybe the user wants both? Not sure of the use case for this.

@tierninho Great question. We discussed this, and creating a new local repo such that they now have both would require somehow moving the in-progress work over from one repo to the other. It's also likely that if they're in this situation, they're an external contributor and therefore having both is likely redundant and potentially confusing.

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    repository.gitHubRepository.htmlURL !== null ? (
+      <>
+        You can try creating the fork manually at{' '}
+        <LinkButton>{repository.gitHubRepository.htmlURL}</LinkButton>.
+      </>
+    ) : (
+      undefined
+    )
+  return (
+    <Dialog
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/forks/create-fork-dialog.tsx]: merged into one component in 5880de4 and 68779dc7d059d3cfc78aac2b45d429b05141d3f8

**Comment by outofambit** [comment]
Hey @niik thanks for all the great feedback!

> Should we perhaps make it clear in the fork dialog that they're forking to their user account?

@ampinsk and I did a pass at the copy to make this more clear. I've included full repo names (owner/repo-name) in the dialog text and updated the screenshots in the PR description (but not the gif).

new copy

> Can we put this behind a beta feature flag for the time being? Just flagging the insertion of the error handler should do the trick.

Good shout! Done in 2c51b27a689c61e7091ab3d0fbbacf8fc5db8c6e.

> Looking at https://developer.github.com/v3/repos/forks/#create-a-fork
> 
> > Note: Forking a Repository happens asynchronously. You may have to wait a short period of time before you can access the git objects. If this takes longer than 5 minutes, be sure to contact GitHub Support.
> 
> Should we perhaps make sure that the repository is all set up and ready for the user before we dismiss the fork dialog?

I have this question too. So far in using the API, I haven't hit a case where the repo isn't ready by the time it needs to be, but its certainly possible. I think I'd like to address this safeguard in a follow up pull request, if that's ok with you?

(My proposal would be some kind of polling with `git ls-remote` to make sure git objects are accessible, with a timeout. Perhaps max 3 seconds poll interval with a 60 second timeout? If it were to timeout but finish later, a subsequent run of the create fork flow would find that fork and use it, not try to make another (this is how the github fork api endpoint works).)

**Comment by outofambit** [comment]
> Supposed to be one dollar sign or two?

😆 i fixed that right after in https://github.com/desktop/desktop/pull/8889/commits/94df80f16b6da00ec2594eb04606cff79ec8695e. Good spot!

**Comment by Rexogamer** [comment]
I’m a bit iffy about the grammar of the error dialog. It just..l doesn’t feel right. Maybe word it “We could not your fork <fork name>”?

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        .
+      </>
+    ) : (
+      undefined
+    )
+  return (
+    <>
+      <DialogContent>
+        <div>
+          {`Making your fork `}
```
Comment (Rexogamer) [inline app/src/ui/forks/create-fork-dialog.tsx]:
```suggestion
          {`We could not make your fork `}
```

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+          <strong>
+            {`${props.account.login}/${props.repository.gitHubRepository.name}`}
+          </strong>
+          {` failed. `}
```
Comment (Rexogamer) [inline app/src/ui/forks/create-fork-dialog.tsx]:
```suggestion
          {` . `}
```

File: app/src/ui/dispatcher/error-handlers.ts [hunk 10]
```diff
@@ -580,6 +585,53 @@ export async function samlReauthRequired(error: Error, dispatcher: Dispatcher) {
...
+  if (!e) {
+    return error
+  }
+
+  const gitError = asGitError(e.underlyingError)
+  if (!gitError || gitError.result.gitError === null) {
+    return error
+  }
+
+  if (!isAuthFailureError(gitError.result.gitError)) {
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/dispatcher/error-handlers.ts]: yes, that's a great point! i looked at working this logic into `AppStore.performPush` at one point, but that method is pretty gnarly as-is. addressed in https://github.com/desktop/desktop/pull/8889/commits/1f1194ffdb837882370dd82e316fa4b23ca312eb

File: app/src/ui/forks/create-fork-dialog.tsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+          {suggestion}
+        </div>
+        <details>
+          <summary>Error details</summary>
+          <pre className="error">{props.error.message}</pre>
```
[the remark this answers is not in the document]
Comment (outofambit) [reply app/src/ui/forks/create-fork-dialog.tsx]: addressed in https://github.com/desktop/desktop/pull/8889/commits/bb6a58ce916c51580eaa4ca2f1586fa2c9acebde

**Comment by outofambit** [comment]
@niik should all be good now!

> Yeah, a follow up is fine. It might not be a problem but we have run into this general type of problem before. I think identifying what the failure mode would be if this happens should be our first priority and if it's not gnarly then we can skip that implementation entirely or wait until we hear it being a problem.

great point. i'm on it! #8949

File: app/src/ui/forks/create-fork-dialog.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (captivemother) [reply app/src/ui/forks/create-fork-dialog.tsx]: _Thank You! I Don't Know What I'm Doing lol, Everyone Told Me It's Easy!!! I'll Try! Thank You Very Much!!!! For The Help._

**Pull request merged**